ui/pages/onboarding-flow/onboarding-flow-switch/onboarding-flow-switch.js

```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import { useSelector } from 'react-redux';
+import { Redirect } from 'react-router-dom';
+import {
+  DEFAULT_ROUTE,
+  LOCK_ROUTE,
+  ONBOARDING_COMPLETION_ROUTE,
+  ONBOARDING_GET_STARTED_ROUTE,
+  ONBOARDING_UNLOCK_ROUTE,
+} from '../../../helpers/constants/routes';
+import {
+  getCompletedOnboarding,
+  getIsInitialized,
+  getIsUnlocked,
+  getSeedPhraseBackedUp,
+} from '../../../selectors';
+
+export default function OnboardingFlowSwitch() {
+  const completedOnboarding = useSelector(getCompletedOnboarding);
+  const isInitialized = useSelector(getIsInitialized);
+  const isUnlocked = useSelector(getIsUnlocked);
+  const seedPhraseBackedUp = useSelector(getSeedPhraseBackedUp);
+
+  if (completedOnboarding) {
+    return <Redirect to={{ pathname: DEFAULT_ROUTE }} />;
+  }
+
+  if (seedPhraseBackedUp !== null) {
+    return <Redirect to={{ pathname: ONBOARDING_COMPLETION_ROUTE }} />;
+  }
+
+  if (isUnlocked) {
+    return <Redirect to={{ pathname: LOCK_ROUTE }} />;
```

Comment:
This seems like a confusing decision :thinking: Why redirect here if the wallet is unlocked?

Reply:
Same thing here. A copy from the original file [here](https://github.com/MetaMask/metamask-extension/blob/6187ab9b01ea757366a42d22b7292413c0f2b119/ui/pages/first-time-flow/first-time-flow-switch/first-time-flow-switch.component.js#L37). Good question 🤔

Reply:
I'm feeling like this if block should be excluded? The default on line 40 is the `onboarding_unlock_route`. Not sure why we would use the normal lock route in the onboarding flow at all since we have a distinct onboarding unlock component. And as you called out the logic seems wrong. Probably supposed to be `if(!isUnlocked)`. But weird that it seems to have been like this in prod for a while... @Gudahtt

Reply:
I swear I've gone down this rabbit hole before, tried to remove this component completely, then found a case where it was required :thinking: I can't recall why though...

Reply:
Oh, more details here: https://github.com/MetaMask/metamask-extension/pull/9825

Reply:
Essentially the lock screen ensures that the onboarding flow works if it gets interrupted. It prevents the seed phrase from being seen without entering the password (if the "create" flow is interrupted), and forcing the unlock has the side-effect of allowing us to bring the seed phrase into the UI for the verification page.

At least for the Create flow it's definitely required! We don't need the _page_, but we need to force the unlock somehow in this circumstance. I'm not totally sure if we need it for the import flow or not :thinking:  But it doesn't seem to hurt at least.

Reply:
Oh, though we'd also need to ensure the Lock screen redirects to the ONBOARDING_V2 unlock page in this case, not the old one.

Reply:
Don't we want a lock screen apart from the onboarding flow? For instance if a user clicks the lock button
<img width="122" alt="Screen Shot 2021-09-30 at 1 00 16 PM" src="https://user-images.githubusercontent.com/34557516/135507247-4e957042-0b19-4adb-9eb2-0db4082a511d.png">
or if they use the auto-lock timer?

Reply:
Not exactly - we don't show a lock screen when that button is used, or when the auto-lock timer is called. We just lock the wallet, and that redirects them to the unlock page.

This "Lock page" isn't really a page, it's a component with an on-mount hook to lock and redirect.

Reply:
And I was wrong there is no new ONBOARDING_V2 unlock page so we should continue to use the existing one. 

I'm still not understanding why [this](https://github.com/MetaMask/metamask-extension/blob/6187ab9b01ea757366a42d22b7292413c0f2b119/ui/pages/first-time-flow/first-time-flow-switch/first-time-flow-switch.component.js#L36) is `if(isUnlock)` and not the inverse?

Reply:
@Gudahtt ^

Reply:
Ok just writing out the onboarding interruption flow so I can think it through.  So the Lock component associated with the `LOCK_ROUTE` actually just sets isUnlocked to false and then history.pushes to DEFAULT_ROUTE. So this component (onboarding-flow-switch, formerly first-time-flow-switch) catches any `onboarding/` routes we don't catch in the routes established in onboarding-flow.js and routes accordingly. If we hit this if block (`if(isUnlocked)`) it means we `completedOnboarding` is false and `seedPhraseBackup` is null and so at this point we want to lock the extension which will push us back to DEFAULT_ROUTE, which because completedOnboarding is false will route us back to ONBOARDING_ROUTE which will then push us to the UNLOCK_ROUTE.

Reply:
Basically yes. But the reason we want to lock the wallet in that case is because we don't want the seed phrase to be shown on the screen without the user verifying that they know the password. If a user got partway through the onboarding flow, closed the page and left the computer, and then a second person came by and opened MetaMask, they'd be shown the seed phrase.

Showing the seed phrase (and similar cases like showing private keys) is the one use case where we try to be careful about shoulder surfers, because of the severe consequences. Hence the "confirm password" step for showing the seed phrase or exporting a private key.

The second reason we wanted to lock the account with the _old_ flow is that the "initializing unlock" screen was a special unlock screen that had a side-effect when it was unlocked: it retrieved the seed phrase itself from the background, and stored it in memory in UI for the seed phrase reveal page and the backup exercise. This is why removing the Lock component caused an infinite loop when the onboarding was interrupted.

I'm not sure how this will work with your new onboarding flow - maybe it makes sense to retrieve it directly when it's needed, rather than store it in state as a side-effect of unlocking on a special onboarding unlock page. So this second reason might not apply anymore, in which case using the normal unlock page might be perfectly fine.

Reply:
Just to clarify in my head, it seems like it's the same `UnlockPage` across the board, but in the initialize/onboarding flow we pass an onSubmit method in that retrieves the seedPhrase and then stores it in state. This still could make sense as the flow. The alternative I'm hearing you describe is to retrieve the seedphrase directly from background on the views where it is needed rather than keeping it in state in the onboarding-flow component and passing it around. The only downside to this second approach, I think, is more calls to the background? Does that sound right to you @Gudahtt?

Reply:
Yes that sounds correct to me. I agree that either seems fine.

> The only downside to this second approach, I think, is more calls to the background

Possibly, though depending on how we do it, we could still guarantee that only one call to the background is made.

Another downside is that having it as a side-effect of unlock _guarantees_ that the user has just entered their password. Whereas if we rely on routing for that, maybe that's easier to mess up.

Reply:
Ok thanks so much as always for your in-depth input @Gudahtt. With regard to this particular issue, I'm thinking if it ain't broke. I believe I've added back what's necessary to preserve this behavior in the new onboarding flow, but perhaps I'm still missing something...?

Reply:
Sorry just noticed that you approved!